modal/parallel_map.py

```diff
@@ -432,3 +562,33 @@ def main():
             "Use Function.map.aio()/Function.for_each.aio() instead."
         ),
     )
+
+
+class _TimestampPriorityQueue:
+    """
+    A priority queue that returns the oldest item that has a timestamp before the current time.
+    """
+
+    _MAX_PRIORITY = float("inf")
+
+    def __init__(self):
+        self._queue = asyncio.PriorityQueue()
+
+    async def close(self):
+        await self._queue.put((self._MAX_PRIORITY, None))
+
+    async def put(self, idx: int, timestamp_seconds: int) -> None:
+        await self._queue.put((timestamp_seconds, idx))
+
+    async def get(self) -> Union[int, None]:
+        while True:
+            (timestamp_seconds, idx) = await self._queue.get()
+            if timestamp_seconds < int(time.time()):
+                return idx
+            if timestamp_seconds == self._MAX_PRIORITY:
+                return None
+            await self._queue.put((timestamp_seconds, idx))
+            await asyncio.sleep(1)
```

Comment:
I think using an `asyncio.Condition` like the previous implementation is better than `asyncio.sleep(1)` since the latter becomes a busy wait.

https://github.com/modal-labs/modal-client/blob/afdc7f8e19758352dda69458cfd7c310fd5eef31/modal/_utils/async_utils.py#L806

Reply:
ah good point. I'll bring back that original implementation.